### PR TITLE
WorkspaceClient Error handling

### PR DESCRIPTION
**Pull request by elioschmutz**
Issuer: #6154 

Dieser PR behandelt die Fehlerbehandlung bei der Kommunikation mit dem Teamraum-Deployment. 

Umsetzungsdetails: https://github.com/4teamwork/opengever.core/issues/6154#issuecomment-584641272

### CODE REVIEW

File: opengever/api/linked_workspaces.py [hunk 1]
```diff
@@ -1,27 +1,98 @@
...
+def request_error_handler(func):
+    """Decorator function to handle request errors to the teamraum deployment.
+
+    We handle HTTP and timeout-errors in a special way. All other request errors
+    are handled by the rest-api with a 500 Internal Server Error.
+    """
+    def handle_http_error(obj, exception):
+        """Handles general http errors.
+
+        We group this errors in a 502 Bad Gateway error.
```
Comment (njohner) [inline opengever/api/linked_workspaces.py]: this -> these

File: opengever/api/linked_workspaces.py [hunk 2]
```diff
@@ -1,27 +1,98 @@
...
+        }
+        return response
+
+    def handle_timeout_error(obj, exception):
+        """Handles all timeout errors. This can be:
+
+        - ReadTimeout
+        - ConnectTimeout
+
+        We group this errors in a 504 Gateway Timeout error.
```
Comment (njohner) [inline opengever/api/linked_workspaces.py]: this -> these

File: opengever/api/tests/test_linked_workspaces.py [hunk 3]
```diff
@@ -146,3 +150,38 @@ def test_raise_exception_for_subdossiers(self, browser):
                 browser.open(
                     subdossier.absolute_url() + '/@linked-workspaces',
                     method='GET', headers={'Accept': 'application/json'}).json
+
+    @browsing
+    def test_request_errors(self, browser):
+        LinkedWorkspacesStorage(self.dossier).add(self.workspace.UID())
+
+        def assertStatusCode(test_with, raises_error):
```
Comment (njohner) [inline opengever/api/tests/test_linked_workspaces.py]: `raises_error` -> `raised_error`

File: opengever/api/tests/test_linked_workspaces.py [hunk 4]
```diff
@@ -146,3 +150,38 @@ def test_raise_exception_for_subdossiers(self, browser):
                 browser.open(
                     subdossier.absolute_url() + '/@linked-workspaces',
                     method='GET', headers={'Accept': 'application/json'}).json
+
+    @browsing
+    def test_request_errors(self, browser):
```
Comment (njohner) [inline opengever/api/tests/test_linked_workspaces.py]: `test_request_error_handling`

File: opengever/api/tests/test_linked_workspaces.py [hunk 5]
```diff
@@ -146,3 +150,38 @@ def test_raise_exception_for_subdossiers(self, browser):
...
+        @contextmanager
+        def teamraum_deployment_error(status_code, text=None):
+            with self.workspace_client_env():
+                browser.login()
+                browser.exception_bubbling = True
+                with requests_mock.Mocker() as mocker:
+                    mocker.post('{}/@@oauth2-token'.format(self.portal.absolute_url()),
+                                status_code=status_code, text=text)
+
+                    with self.assertRaises(HTTPServerError):
```
Comment (njohner) [inline opengever/api/tests/test_linked_workspaces.py]: I'd move that assertion into the `assertStatusCode` method, it's a bit strange that in the context manager we make that assertion.

File: opengever/api/tests/test_linked_workspaces.py [hunk 6]
```diff
@@ -146,3 +150,38 @@ def test_raise_exception_for_subdossiers(self, browser):
...
+                            self.dossier.absolute_url() + '/@linked-workspaces',
+                            method='GET', headers={'Accept': 'application/json'}).json
+
+                    yield browser
+
+        assertStatusCode(test_with=404, raises_error=502)
+        assertStatusCode(test_with=408, raises_error=504)
+        assertStatusCode(test_with=500, raises_error=502)
+        assertStatusCode(test_with=502, raises_error=502)
+        assertStatusCode(test_with=504, raises_error=504)
```
Comment (njohner) [inline opengever/api/tests/test_linked_workspaces.py]: I find the tests a bit hard to read. Maybe I'd get rid of the `teamraum_deployment_error` context manager?

File: opengever/api/linked_workspaces.py (hunk 1 above)
Comment (elioschmutz) [reply opengever/api/linked_workspaces.py]: 🤦‍♂ how many times do I have to make this mistake until I'll learn it?

File: opengever/api/tests/test_linked_workspaces.py (hunk 6 above)
Comment (elioschmutz) [reply opengever/api/tests/test_linked_workspaces.py]: 👍 I removed the contextmanager

**Pull request merged**